pyModeS/extra/tcpclient.py

```diff
@@ -149,6 +151,98 @@ def read_beast_buffer(self):
         return messages
 
+    
+    def read_beast_buffer_rssi(self):
+        """Handle mode-s beast data type.
+
+        <esc> "1" : 6 byte MLAT timestamp, 1 byte signal level,
+            2 byte Mode-AC
+        <esc> "2" : 6 byte MLAT timestamp, 1 byte signal level,
+            7 byte Mode-S short frame
+        <esc> "3" : 6 byte MLAT timestamp, 1 byte signal level,
+            14 byte Mode-S long frame
+        <esc> "4" : 6 byte MLAT timestamp, status data, DIP switch
+            configuration settings (not on Mode-S Beast classic)
+        <esc><esc>: true 0x1a
+        <esc> is 0x1a, and "1", "2" and "3" are 0x31, 0x32 and 0x33
+
+        timestamp:
+        wiki.modesbeast.com/Radarcape:Firmware_Versions#The_GPS_timestamp
+        """
+        messages_mlat = []
+        msg = []
+        i = 0
+
+        # process the buffer until the last divider <esc> 0x1a
+        # then, reset the self.buffer with the remainder
+
+        while i < len(self.buffer):
+            if self.buffer[i : i + 2] == [0x1A, 0x1A]:
+                msg.append(0x1A)
+                i += 1
+            elif (i == len(self.buffer) - 1) and (self.buffer[i] == 0x1A):
+                # special case where the last bit is 0x1a
+                msg.append(0x1A)
+            elif self.buffer[i] == 0x1A:
+                if i == len(self.buffer) - 1:
+                    # special case where the last bit is 0x1a
+                    msg.append(0x1A)
+                elif len(msg) > 0:
+                    messages_mlat.append(msg)
+                    msg = []
+            else:
+                msg.append(self.buffer[i])
+            i += 1
+
+        # save the reminder for next reading cycle, if not empty
+        if len(msg) > 0:
+            reminder = []
+            for i, m in enumerate(msg):
+                if (m == 0x1A) and (i < len(msg) - 1):
+                    # rewind 0x1a, except when it is at the last bit
+                    reminder.extend([m, m])
+                else:
+                    reminder.append(m)
+            self.buffer = [0x1A] + msg
+        else:
+            self.buffer = []
+
+        # extract messages
+        messages = []
+        for mm in messages_mlat:
+            ts = time.time()
+
+            msgtype = mm[0]
+
+            if msgtype == 0x32:
+                # Mode-S Short Message, 7 byte, 14-len hexstr
+                msg = "".join("%02X" % i for i in mm[8:15])
+            elif msgtype == 0x33:
+                # Mode-S Long Message, 14 byte, 28-len hexstr
+                msg = "".join("%02X" % i for i in mm[8:22])
+            else:
+                # Other message type
+                continue
+
+            if len(msg) not in [14, 28]:
+                continue
+
+            #calculate linear power value
+            df = pms.df(msg)
+            raw_rssi = mm[7]
+            voltage = raw_rssi / 255
+            power = voltage ** 2 
+            rssi = 10 * math.log10(power) 
```

Comment:
This is often referred to online as dBm, and is opposed to PmW (what you name power). But you seem to assume that the power is that percentage squared??

Reply:
Let me clear up my methodology. My company uses a device called PiAware which emits RSSI values as linear voltage measurements. This means that each measurement is essentially sqrt(power) * 255 (as power is proportional to voltage squared). My company wanted me to convert this value to a dBFS power value. To do so:

1) we get the raw 0-255 byte value (raw_rssi = mm[7])
2) we scale it to 0.0 - 1.0 (voltage = raw_rssi / 255)
3) we convert it to a dBFS power value (rolling the squaring of the voltage into the dB calculation) 

Essentially, we have performed the equation 20log10(signal/255). It's important to note that we used dBFS (a relative unit of measurement) compared to dBm (an absolute unit of measurement).

Reply:
Thanks for the explanation. I am not very familiar with these aspects. I am just wondering whether those variable names are appropriate. Or what would be relevant comments in the code.

From few references I checked online, `rffi` is often defined as the percentage that you compute and named `voltage` (but then for me there's a dimension problem if we call that voltage, no?) Also maybe name the final value `dbfs`?

The question is also whether we are following the beast format standard or just reverse engineering this:

https://github.com/flightaware/dump1090/blob/a80ba8f82a74c90a29619ddbc10909c561198541/net_io.c#L473

So what you call `voltage` is the `signalLevel` in the code?

But that `signalLevel` is computed as the average of the magnitude of the I/Q samples along the message divided by 65535 * 65535 (assuming you get the samples as int16*int16, makes sense). So why voltage? That's really just a ratio (between 0 and 1, further encoded as an int8).

Reference:
https://github.com/flightaware/dump1090/blob/a80ba8f82a74c90a29619ddbc10909c561198541/demod_2400.c#L381

Also, it would probably make sense to return None when the value is 0xff, according to the standard and to the implementation there.

By the way, do you understand that implementation here? Why the sum??

https://github.com/flightaware/dump1090/blob/a80ba8f82a74c90a29619ddbc10909c561198541/net_io.c#L1859

Reply:
(That's really thinking aloud, I am only trying to understand... Sorry if that is annoying)

Reply:
I'm glad you asked these questions. I agree with a lot of the changes you suggest concerning renaming the variables and changing the comments to be more descriptive and plan to make them in my next commit. I consulted the developer of the dump1090 library (Oliver Jowett) for this project and this is the methodology he proposed. I can ask him the questions you raised. I feel in some ways that I'm reverse engineering the methods used by dump1090. I might need to figure out a way that works for all devices rather than just those used by dump1090.

Reply:
Thanks for taking it positively. Maybe we keep it as is in the end, but at least we will know why :)